pkg/controller/drain/drain_controller.go

```diff
@@ -0,0 +1,417 @@
+	for ctrl.processNextWorkItem() {
+	}
+}
+
+func (ctrl *Controller) processNextWorkItem() bool {
+	key, quit := ctrl.queue.Get()
+	if quit {
+		return false
+	}
+	defer ctrl.queue.Done(key)
+
+	err := ctrl.syncHandler(key.(string))
+	ctrl.handleErr(err, key)
+
+	return true
+}
+
+func (ctrl *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		ctrl.queue.Forget(key)
+		return
+	}
+
+	if ctrl.queue.NumRequeues(key) < maxRetries {
+		glog.V(2).Infof("Error syncing node %v: %v", key, err)
+		ctrl.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	glog.V(2).Infof("Dropping node %q out of the queue: %v", key, err)
+	ctrl.queue.Forget(key)
+	ctrl.queue.AddAfter(key, 1*time.Minute)
+}
+
+func (ctrl *Controller) syncNode(key string) error {
+	startTime := time.Now()
+	glog.V(4).Infof("Started syncing node %q (%v)", key, startTime)
+	defer func() {
+		glog.V(4).Infof("Finished syncing node %q (%v)", key, time.Since(startTime))
+	}()
+
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	// First check if a drain request is needed for this node
+	// TODO (jerzhang): consider not using the cache for this,
+	// since it has been shown to be problematic in the past
+	node, err := ctrl.nodeLister.Get(name)
+	if apierrors.IsNotFound(err) {
+		glog.V(2).Infof("node %v has been deleted", key)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	desiredState := node.Annotations[daemonconsts.DesiredDrainerAnnotationKey]
+	if desiredState == node.Annotations[daemonconsts.LastAppliedDrainerAnnotationKey] {
+		glog.V(4).Infof("Node %v has the correct drain", key)
+		return nil
+	}
+
+	podName, ok := os.LookupEnv("POD_NAME")
+	if !ok || podName == "" {
+		glog.Fatalf("pod name is required for controller drainer")
+	}
+
+	drainer := &drain.Helper{
+		Client:              ctrl.kubeClient,
+		Force:               true,
+		IgnoreAllDaemonSets: true,
+		DeleteEmptyDirData:  true,
+		GracePeriodSeconds:  -1,
+		Timeout:             90 * time.Second,
+		OnPodDeletedOrEvicted: func(pod *corev1.Pod, usingEviction bool) {
+			verbStr := "Deleted"
+			if usingEviction {
+				verbStr = "Evicted"
+			}
+			glog.Infof("%s pod %s/%s", verbStr, pod.Namespace, pod.Name)
+		},
+		AdditionalFilters: []drain.PodFilter{
+			func(pod corev1.Pod) drain.PodDeleteStatus {
+				if pod.Name == podName {
+					return drain.MakePodDeleteStatusSkip()
+				}
+				return drain.MakePodDeleteStatusOkay()
+			},
+		},
+		Out:    writer{glog.Info},
+		ErrOut: writer{glog.Error},
+		Ctx:    context.TODO(),
+	}
+
+	desiredVerb := strings.Split(desiredState, "-")[0]
+	if desiredVerb == daemonconsts.DrainerStateUncordon {
+		glog.Infof("node %v to be uncordoned", key)
+		// perform uncordon
+		if err := ctrl.cordonOrUncordonNode(false, node, drainer); err != nil {
+			return fmt.Errorf("failed to uncordon node %v: %w", node.Name, err)
+		}
+	}
+
+	if desiredVerb == daemonconsts.DrainerStateDrain {
```

Comment:
This could be an `else if` or a `switch`.

Reply:
I'll switch to a switch :)